Improve semantic for packages:all:require

alalazo opened the pull request:
This PR improves the semantic for `packages:all:require` by rejecting constraints that would lead to clearly unsolvable cases. The semantic of package specific requirements stays the same as before.

The improvement is that users can write:
```yaml
packages:
  all:
    require:
    - "+shared"
    - "+mpi"
``` 
and the constraints will be enforced only for packages that could support them.  This means that for a package like `zlib`, which only has the `shared` variant, then only `+shared` will be enforced.

Constraints are also always considered as a whole, and never partially enforced. For instance, the difference between the configuration above and:
```yaml
packages:
  all:
    require: "+mpi +shared"
``` 
is that, in the latter case, the requirement is enforced only on packages that have both the `mpi` and the `shared` variant.

There is a distinction between type of variants, so a boolean variant will never match a multi-valued variant with the same name and vice-versa.

Modifications:
- [x] Improved the semantic of `packages:all:require`
- [x] Added unit-tests and documentation

alalazo committed:
requirements: add unit-test to specify semantic of 'all:'

alalazo committed:
requirements: dedent an "else" branch

alalazo committed:
Improve semantic for packages:all:require

A requirement is emitted for the solver only if
the corresponding constraint is solvable for
the package. Otherwise, the constraint is rejected.

alalazo committed:
Added documentation

alalazo commented:
@albestro

albestro commented:
Just tested and it works nicely as expected now! Thanks a lot!

### BEFORE

```console
$ spack -c "packages:all:require:+cuda" spec zlib
==> Error: concretization failed for the following reasons:

   1. cannot satisfy a requirement for package 'zlib'.
   2. cannot satisfy a requirement for package 'gmake'.
```

```console
$ spack -c "packages:all:require:[+cuda,~shared,~pic]" spec zlib
==> Error: concretization failed for the following reasons:

   1. cannot satisfy a requirement for package 'zlib'.
   2. cannot satisfy a requirement for package 'gmake'.
```

### AFTER

```console
$ spack -c "packages:all:require:+cuda" spec zlib
Input spec
--------------------------------
 -   zlib

Concretized
--------------------------------
 -   zlib@1.3%gcc@11.2.0+optimize+pic+shared build_system=makefile arch=cray-cnl7-haswell
[+]      ^gmake@4.4.1%gcc@11.2.0~guile build_system=autotools arch=cray-cnl7-haswell
```

```console
$ spack -c "packages:all:require:[+cuda,~shared,~pic]" spec zlib
Input spec
--------------------------------
 -   zlib

Concretized
--------------------------------
 -   zlib@1.3%gcc@11.2.0+optimize~pic~shared build_system=makefile arch=cray-cnl7-haswell
[+]      ^gmake@4.4.1%gcc@11.2.0~guile build_system=autotools arch=cray-cnl7-haswell
```

haampie commented on lib/spack/docs/packages_yaml.rst:
```diff
@@ -383,20 +383,52 @@ like this:
 which means every spec will be required to use ``clang`` as a compiler.
 
-Note that in this case ``all`` represents a *default set of requirements* -
+Any constraint that appears in default requirements under ``all``, and that
+would clearly lead to unsolvable cases, is disregarded. For instance a configuration
+like the following:
```
```suggestion
Requirements on variants for all packages are possible too, but note that they
are only enforced for those packages that define these variants, otherwise they
are disregarded. For example:
```

I think it's important to stress "variants" here, cause there are other constraints that can "clearly" lead to unsolvable cases like `@1:`.

alalazo committed:
Address review comment on docs

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1338,6 +1352,26 @@ def _rule_from_str(
+        self, pkg_name: str, *, constraint: str, kind: RequirementKind
+    ) -> bool:
+        """Returns True if a requirement constraint should be rejected"""
+        if kind == RequirementKind.DEFAULT:
```
Maybe never call this if `kind != DEFAULT` and drop the `kind` argument?

Edit: maybe a bit too much duplication to move the branch out of the loop.

alalazo replied:
Hmm, I thought about that, but then I'll just have to push the check on `kind` to the caller in three places, and disconnect it from the check that the variants, compilers, versions etc. exists.

If you worry about a change in API I can add a leading `_`.

haampie replied:
Yeah, I don't feel so strongly about it, but naturally I would move a static branch out of the loop. But probably the loop is not "hot" enough to make it worth the effort.

haampie replied:
```python
    def requirement_rules_from_packages_yaml(self, pkg):
        pkg_name = pkg.name
        config = spack.config.get("packages")
        requirements = config.get(pkg_name, {}).get("require", [])
        if requirements:
            kind = RequirementKind.PACKAGE
            drop = lambda requirement: False
        else:
            requirements = config.get("all", {}).get("require", [])
            kind = RequirementKind.DEFAULT
            drop = lambda requirement: self.reject_requirement_constraint(pkg_name, constraint=requirement)
        return self._rules_from_requirements(pkg_name, requirements, drop=drop, kind=kind)
```
?

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1296,34 +1296,48 @@ def _rules_from_requirements(self, pkg_name: str, requirements, *, kind: Require
+                rules.append(self._rule_from_str(pkg_name, requirement, kind))
+            else:
+                for policy in ("spec", "one_of", "any_of"):
+                    if policy in requirement:
+                        constraints = requirement[policy]
```
can you also de-indent?

alalazo replied:
Yes. Being in a loop I wasn't sure it was more readable to other people too. Since you ask, I'll go for that.

alalazo committed:
Dedent for loop, invert loop condition

alalazo committed:
Simplify encoding for strings

alalazo committed:
Simplify treatment of strings

Transform internally to the general object and then
have a single flow for the code
